Team417_CENTERSTAGE/src/main/java/org/firstinspires/ftc/team417_CENTERSTAGE/BaseTeleOp.java

```diff
@@ -0,0 +1,249 @@
+            for (int i = 0; i < backupArmPositions.length; i++)
+            {
+                if (backupArmPositions[i] > armLocation) {
+                    armLocation = backupArmPositions[i];
+                    break;
+                }
+            }
+        } else if (gamepad2.dpad_down && !dpadDownPressed){ //if dpad down is being pressed change the arm location to the last location in the arm positions array.
+            for (int i = backupArmPositions.length - 1; i >= 0; i--)
+            {
+                if (backupArmPositions[i] > armLocation) {
+                    armLocation = backupArmPositions[i];
+                    break;
+                }
+            }
+        }
+
+        dpadDownPressed = gamepad2.dpad_down;
+        dpadUpPressed = gamepad2.dpad_up;
+
+
+        if (armMotor.getCurrentPosition() < armLocation + 50 && armMotor.getCurrentPosition() > armLocation - 50) {
+            armMotor.setPower(0);
+        } else if (armMotor.getCurrentPosition() > armLocation) {
+            armMotor.setPower(-0.7);
+        } else if (armMotor.getCurrentPosition() < armLocation) {
+            armMotor.setPower(0.7);
+        }
+
+        telemetry.addData("arm target position", armLocation);
+        //armMotor.setTargetPosition((int) Math.round(armLocation)); //tell the motor to move to arm location.
+        //armMotor.setMode(DcMotor.RunMode.RUN_TO_POSITION);
+    }
+
+    public double armPositionIndex;
+
+    public void controlArmUsingControllers() {
+        moveArm(-gamepad2.right_stick_y);
+
+        for (int i = 0; i < armPositions.length; i++) {
+            if (armMotor.getCurrentPosition() < armPositions[i]) {
+                armPositionIndex = i - 0.5;
+            } else if (isEpsilonEquals(armMotor.getCurrentPosition(), armPositions[i])) {
+                armPositionIndex = i;
+            }
+        }
+
+        if (isEpsilonEquals(armPositionIndex % 1.0, 0) || isEpsilonEquals(armPositionIndex % 1.0, 1)) {
+            if (gamepad2.dpad_up && gamepad2.dpad_down) {
+            } else if (gamepad2.dpad_up) {
+                positionArm(((int) Math.round(armPositionIndex)) + 1);
+            } else if (gamepad2.dpad_down) {
+                positionArm(((int) Math.round(armPositionIndex)) - 1);
+            }
+        } else {
+            if (gamepad2.dpad_up && gamepad2.dpad_down) {
+            } else if (gamepad2.dpad_up) {
+                positionArm((int) Math.ceil(armPositionIndex));
+            } else if (gamepad2.dpad_down) {
+                positionArm((int) Math.floor(armPositionIndex));
+            }
+        }
+    }
+
+    public boolean dumperTilted = false;
+    public boolean xIsPressed = false;
+    public boolean dumperDumped = false;
+    public boolean bIsPressed = false;
+
+    public void controlDumperUsingControllers() {
+        /*
+        double dumperSpeed = -gamepad2.right_stick_y;
```

Comment:
Why is this code commented out? It would have been a good idea to have left a comment why.